### PR TITLE
cohttp-eio: add cochttp-eio based benchmark

### PR DESCRIPTION
**Pull request by bikallem**
Change summary:
1. Creates a dockerfile for each web server so that we can individually build/run benchmark as required. 
2. Add Makefile target `build` and `run` to to build and run separately
3. Add multi core benchmark script separate from 1 core. 
4. Add multi core rust benchmark.

### CODE REVIEW

**Review by talex5 — CHANGES_REQUESTED** [review]
Since this implementation is giving impossible results, some changes are obviously needed.

I suggest writing a validator program that behaves a bit like wrk2 (i.e. sends lots of requests in parallel) but ensures that it's getting correct responses. In particular:

1. Make sure it gets exactly the expected reply to each request. No truncated or extended versions.
2. Make sure it never gets more replies than it asked for.
3. Check that all connections are being serviced. It's quite easy to game the benchmark by e.g. just sending replies on 10 connections really fast, while stalling all other connections.

That should show what's wrong with your code, and would also be a really useful check that the others are working correctly (before then testing their speed with `wrk2`, as before).

**Comment by bikallem** [comment]
With regards to the correctness of cohttp-eio, here is a small demo of it working. The response size, and the response in general seems to be correct. Below shows the server working along side the browser. 
 
cohttp-eio

~~After going through some literature regarding wrk2, it seems the wrk2 methodology is quite solid and tested in the field. At least the arguments for hdr histrogram and coordinated omissions sound very cogent. The response serviced/second is a sampling value, so I am inclined to think this is also correct(?). Perhaps https://github.com/gs0510 can confirm.~~

What puzzles me most is the steep drop in/around 150k/sec requests? Is there a memory space contention perhaps between the load generator and the server?

References:
1. https://github.com/giltene/wrk2 - Readme is quite comprehensive
2. https://www.quora.com/In-Java-what-is-Coordinated-Omission 
3. http://hdrhistogram.org - central to wrk2 methodology I believe

**Comment by art-w** [comment]
So, this was "interesting". If you look at wrk2 log `output/run-cohttp_eio.exe-1000-1000.txt`:

```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    35.43s    14.26s    1.00m    57.73%
    Req/Sec    10.17k   390.77    10.63k    48.00%
...
  14978619 requests in 1.00m, 30.63GB read
```

How can wrk2 perform 15 million requests when it is scheduled to run a thousand per second, and how can the requests have an average of 35s latency?

Well, it turns out that wrk2 throttles the requests when *it* send them... but it doesn't throttle when reading the response from the server. As wrk2 is running in keep alive mode, the server is allowed to respond multiple times and wrk2 counts each additional response as a completed "request". However, wrk2 measures the response latency from the moment the initial http query was sent, so each new response from the server has an increasing latency (from 0ms for the first response to 1min for the last, yielding a large avg).

So yeah in practice, wrk2 sends 1'000 initial requests... and then get flooded with a never ending stream of responses.

I haven't had time to look into cohttp-eio, but you can reproduce the infinite stream of messages with:

```
$ # echo -e "GET / HTTP/1.1\r\nHost: localhost\r\n\r\n" | nc localhost 8080
HTTP/1.1 200 OK
Date: Thu, 03 Feb 2022 00:04:15 GMT
content-length: 2053
connection: keep-alive
content-type: text/plain; charset=UTF-8

CHAPTER I. Down the Rabbit-Hole  ......HTTP/1.1 200 OK
Date: Thu, 03 Feb 2022 00:04:15 GMT
content-length: 2053
connection: keep-alive
content-type: text/plain; charset=UTF-8

CHAPTER I. Down the Rabbit-Hole ......HTTP/1.1 200 OK
Date: Thu, 03 Feb 2022 00:04:15 GMT
content-length: 2053
connection: keep-alive
content-type: text/plain; charset=UTF-8

...
```

**Comment by bikallem** [comment]
> I haven't had time to look into cohttp-eio, but you can reproduce the infinite stream of messages with:

I think I have pinpointed the issue. I will re-run the benchmark and post updates.

**Comment by bikallem** [comment]
This is ready now.

**Comment by bikallem** [comment]
@talex5 This is ready to merge.

File: rust-hyper/src/main.rs [hunk 1]
```diff
@@ -2,54 +2,98 @@ extern crate futures;
 extern crate hyper;
 extern crate net2;
 extern crate num_cpus;
+#[macro_use]
+extern crate serde_derive;
+extern crate serde_json;
 extern crate tokio_core;
 
-/*
-This is adapted from the Techempower Hyper benchmark: https://github.com/TechEmpower/FrameworkBenchmarks/tree/master/frameworks/Rust/hyper
```
Comment (talex5) [inline rust-hyper/src/main.rs:8]: This seems like useful information - why was it removed?

**Review by talex5 — COMMENTED** [review]
Thanks - this looks reasonable now. What were the final results?

**Review by talex5 — APPROVED** [review]

File: rust-hyper/src/main.rs (hunk 1 above)
Comment (bikallem) [reply rust-hyper/src/main.rs:8]: I think it got removed during code shuffling. Added again.

**Comment by bikallem** [comment]
Should I merge this if there aren't any objections?

**Pull request merged**